fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java

```diff
@@ -0,0 +1,190 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.cursors;
+
+import com.apple.foundationdb.API;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.RecordCursorResult;
+import com.apple.foundationdb.record.RecordCursorVisitor;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunner;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.BiFunction;
+
+/**
+ * A cursor that can iterate over a cursor across transactions.
+ *
+ * It is provided a generator that produces a cursor (referred to as <i>underlying cursor</i>). The <i>underlying
+ * cursor</i> is iterated over until it is either exhausted or until one of the scan limits is reached:
+ * <ul>
+ *     <li>
+ *         If the <i>underlying cursor</i> is exhausted, the {@link AutoContinuingCursor} is also exhausted.
+ *     </li>
+ *     <li>
+ *         If scan limit properties of the <i>underlying cursor</i> are reached, the generator is asked for a new
+ *         <i>underlying cursor</i> which (1) is in the context of a new transaction and (2) takes the continuation from
+ *         the previous <i>underlying cursor</i>. Then the process is repeated.
+ *     </li>
+ * </ul>
+ *
+ * <p>
+ * {@link AutoContinuingCursor} is responsible for all {@link FDBRecordContext} management, so all reads it does are in
+ * the scope of transactions that it controls.
+ * </p>
+ *
+ * <p>
+ * The {@link AutoContinuingCursor} has no visibility into the {@link com.apple.foundationdb.record.ScanProperties} of
+ * the <i>underlying cursor</i> and, therefore, will not be involved in enforcing any limits that may be individually
+ * applied to the <i>underlying cursor</i>. For example, if the generator returns an <i>underlying cursor</i> that
+ * specified, say, a record scan limit of 10 records, the {@link AutoContinuingCursor} will scan all data until it is
+ * exhausted, at most 10 records at a transaction.
+ * </p>
+ *
+ * @param <T> the type of elements returned by this cursor
+ */
+@API(API.Status.EXPERIMENTAL)
+public class AutoContinuingCursor<T> implements BaseCursor<T> {
```

Comment:
Since the `AutoContinuingCursor` has a child and doesn't enforce the scan limits, I don't think it should be a `BaseCursor`. It should probably just implement `RecordCursor<T>` instead.

Reply:
Done.